### PR TITLE
Fix kwargs for QuantizationAwareTraining callback

### LINKED ISSUE #7926: QAT callback doesn't work with multi-input models or forward kwargs
## 🐛 Bug

The `QuantizationAwareTraining` can only handle models whose forward signature looks like `forward(data)`. Adding additional positional or keyword arguments breaks the forward propagation.

## Please reproduce using the BoringModel

The `BoringModel` doesn't have the right forward signature to produce errors, but I included an equivalently simple model below

### To Reproduce

```python
import functools
import torch
import torch.nn as nn


class Simple(nn.Module):
    def __init__(self):
        super(Simple, self).__init__()
        self.conv = nn.Conv2d(3, 16, kernel_size=3)
        self.softmax = nn.Softmax(dim=3)

    def forward(self, a, b, softmax=False):
        """Use 2 positional arguments and 1 keyword argument"""
        feat_b = self.conv(b)
        feat_a = self.conv(a)
        combined = feat_a + feat_b
        if softmax:
            return self.softmax(combined)
        else:
            return combined

    @property
    def example_input_array(self):
        return torch.randn(1, 3, 28, 28), torch.randn(1, 3, 28, 28)


def wrap_quantize_forward_context(model, func):
    """From QuantizationAwareTraining callback"""
    @functools.wraps(func)
    def wrapper(data):
        data = model.quant(data)
        data = func(data)
        data = model.dequant(data)
        return data
    return wrapper


# Works fine without quantization
model = Simple()
input_sample = model.example_input_array
_ = model(*input_sample)
_ = model(*input_sample, softmax=True)

# Simulate the `on_fit_start` hook behavior
model.quant = torch.quantization.QuantStub()
model.dequant = torch.quantization.DeQuantStub()
model.forward = wrap_quantize_forward_context(model, model.forward)
_ = model(*input_sample)
_ = model(*input_sample, softmax=True)
```

The last two lines will raise the following exceptions

```
TypeError: wrapper() takes 1 positional argument but 2 were given
TypeError: wrapper() got an unexpected keyword argument 'softmax'
```

### Expected behavior

Forward prop should work fine for both the quantized and unquantized model

### Environment

* CUDA:
        - GPU: GeForce GTX 1060 with Max-Q Design
        - available:         True
        - version:           10.2
* Packages:
        - numpy:             1.19.2
        - pyTorch_debug:     False
        - pyTorch_version:   1.8.1+cu102
        - pytorch-lightning: 1.3.0
        - tqdm:              4.59.0
* System:
        - OS:                Linux
        - architecture: 64bit
        - processor:         x86_64
        - python:            3.6.13
        - version:           #148-Ubuntu SMP Sat May 8 02:33:43 UTC 2021

### Additional context

The solution to the minimal example is to make the decoration wrapper more generic

```python
def wrap_quantize_forward_context(model, func):
    """From QuantizationAwareTraining callback"""
    @functools.wraps(func)
    def wrapper(*data, **kwargs):
        data = model.quant(data)
        data = func(*data, **kwargs)
        data = model.dequant(data)
        return data
    return wrapper
```

However, I notice that `callbacks/quantization.py` also includes a second decorator `wrap_qat_forward_context`, and I am not exactly clear on the intent behind having two separate decorators. I see that `wrap_quantize_forward_context` is only used in the `on_fit_end` hook. Any insight here @Borda?

I can make a PR for the above solution, but first I tested it in an actual training by patching my local install

```python
def wrap_qat_forward_context(
    quant_cb,
    model: 'pl.LightningModule',
    func: Callable,
    trigger_condition: Optional[Union[Callable, int]] = None
) -> Callable:
    @functools.wraps(func)
    def wrapper(*data, **kwargs) -> Any:  # <-- Changes to decoration signature

        _is_func_true = isinstance(trigger_condition, Callable) and trigger_condition(model.trainer)
        _is_count_true = isinstance(trigger_condition, int) and quant_cb._forward_calls < trigger_condition
        _quant_run = trigger_condition is None or _is_func_true or _is_count_true
        if _quant_run:
            quant_cb._forward_calls += 1
            data = model.quant(data)  # <-- Triggers AttributeError 'tuple' object has no attribute 'detach'
        data = func(*data, **kwargs)
        if _quant_run:
            data = model.dequant(data)
        return data

    return wrapper
```

So I tried also expanding the `data` before passing to `model.quant` but this goes back to a `TypeError`

```python
def wrap_qat_forward_context(
    quant_cb,
    model: 'pl.LightningModule',
    func: Callable,
    trigger_condition: Optional[Union[Callable, int]] = None
) -> Callable:
    @functools.wraps(func)
    def wrapper(*data, **kwargs) -> Any:

        _is_func_true = isinstance(trigger_condition, Callable) and trigger_condition(model.trainer)
        _is_count_true = isinstance(trigger_condition, int) and quant_cb._forward_calls < trigger_condition
        _quant_run = trigger_condition is None or _is_func_true or _is_count_true
        if _quant_run:
            quant_cb._forward_calls += 1
            data = model.quant(*data)  # <-- Attempted tuple unpacking, get TypeError: forward() takes 2 positional arguments but 3 were given
        data = func(*data, **kwargs)
        if _quant_run:
            data = model.dequant(*data)
        return data

    return wrapper
```

### PR DESCRIPTION
**Pull request by addisonklinke**
## What does this PR do?

- Supports `*args, **kwargs` syntax in `QuantizationAwareTraining` forward wrappers
- Fixes #7926

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by addisonklinke** [comment]
I'm not able to view exactly where the codecov diff was introduced, but would be happy to work on improving the coverage with some guidance

File: CHANGELOG.md [hunk 1]
```diff
@@ -164,6 +165,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 - Added `on_load_checkpoint` and `on_save_checkpoint` hooks to the `PrecisionPlugin` base class ([#7831](https://github.com/PyTorchLightning/pytorch-lightning/pull/7831))
 
+- `QuantizationAwareTraining` callback requires either `max_num_inputs` parameter or `model.example_input_array` in order to determine the required number of `QuantStubs` ([#7979](https://github.com/PyTorchLightning/pytorch-lightning/pull/7979))
```
Comment (carmocca) [inline CHANGELOG.md:168]: Should have 2 spaces between CHANGELOG items

```suggestion

- `QuantizationAwareTraining` callback requires either `max_num_inputs` parameter or `model.example_input_array` in order to determine the required number of `QuantStubs` ([#7979](https://github.com/PyTorchLightning/pytorch-lightning/pull/7979))

```

File: CHANGELOG.md [hunk 2]
```diff
@@ -79,9 +79,10 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 - Added LightningCLI support for argument links applied on instantiation ([#7895](https://github.com/PyTorchLightning/pytorch-lightning/pull/7895))
 
+- Added support for quantization of multi-input models and forward method kwargs ([#7979](https://github.com/PyTorchLightning/pytorch-lightning/pull/7979))
 
-### Changed
 
+### Changed
```
Comment (carmocca) [inline CHANGELOG.md:85]:
```suggestion
### Changed

```

File: pytorch_lightning/callbacks/quantization.py [hunk 3]
```diff
@@ -114,6 +117,10 @@ def __init__(
             observer_type: allows switching between ``MovingAverageMinMaxObserver`` as "average" (default)
                 and ``HistogramObserver`` as "histogram" which is more computationally expensive.
 
+            max_num_inputs: maximum number of inputs nodes used by the model in training, evaluation, or
+                testing. This determines how many ``QuantStubs`` are needed. If not specified, the model's
```
Comment (carmocca) [inline pytorch_lightning/callbacks/quantization.py:121]:
```suggestion
                testing. This determines how many :class:`torch.quantization.QuantStub`s are needed. If not specified, the model's
```

File: pytorch_lightning/callbacks/quantization.py [hunk 4]
```diff
@@ -114,6 +117,10 @@ def __init__(
...
+                ``example_input_array`` property must be defined
```
Comment (carmocca) [inline pytorch_lightning/callbacks/quantization.py:122]: might need some adjusting

```suggestion
                :attr:~pytorch_lightning.core.lightning.LightningModule.example_input_array` property must be defined
```

File: pytorch_lightning/callbacks/quantization.py [hunk 5]
```diff
@@ -174,7 +182,15 @@ def _check_feasible_fuse(self, model):
...
+            if pl_module.example_input_array is None:
+                raise MisconfigurationException(
+                    '`max_num_inputs` or `model.example_input_array must be provided to determine number of QuantStubs'
+                )
+            num_inps = len(pl_module.example_input_array)
+        else:
+            num_inps = self.max_num_inputs
+        pl_module.quants = nn.ModuleList([torch.quantization.QuantStub() for _ in range(num_inps)])
         # DeQuantStub converts tensors from quantized to floating point
         pl_module.dequant = torch.quantization.DeQuantStub()
```
Comment (carmocca) [inline pytorch_lightning/callbacks/quantization.py:195]: noob question: do we not need a list for these?

File: pytorch_lightning/callbacks/quantization.py [hunk 6]
```diff
@@ -174,7 +182,15 @@ def _check_feasible_fuse(self, model):
...
-        pl_module.quant = torch.quantization.QuantStub()
+        if self.max_num_inputs is None:
+            if pl_module.example_input_array is None:
+                raise MisconfigurationException(
+                    '`max_num_inputs` or `model.example_input_array must be provided to determine number of QuantStubs'
+                )
+            num_inps = len(pl_module.example_input_array)
+        else:
+            num_inps = self.max_num_inputs
+        pl_module.quants = nn.ModuleList([torch.quantization.QuantStub() for _ in range(num_inps)])
```
Comment (carmocca) [inline pytorch_lightning/callbacks/quantization.py:193]: This means it will fail for previous users who don't update right?

Maybe the previous behaviour should be kept if none are provided

**Review by carmocca — COMMENTED** [review]
> I'm not able to view exactly where the codecov diff was introduced, but would be happy to work on improving the coverage with some guidance

Don't sweat this too much. The codecov bot/report sometimes goes crazy considering we run many jobs and the report is not updated properly until the end

File: pytorch_lightning/callbacks/quantization.py (hunk 5 above)
Comment (addisonklinke) [reply pytorch_lightning/callbacks/quantization.py:195]: `nn.ModuleList` is preferable to plain list since it makes PyTorch aware that the `QuantStubs` should be including in `self.parameters()`

File: pytorch_lightning/callbacks/quantization.py (hunk 6 above)
Comment (addisonklinke) [reply pytorch_lightning/callbacks/quantization.py:193]: Good point, I'll update so that a single `QuantStub` is used if neither `max_num_inputs` or `example_input_array` is provided. Then there won't be any breaking changes. Maybe a user warning would be good though? The traceback for multi-input models is not super intuitive

File: pytorch_lightning/callbacks/quantization.py (hunk 5 above)
Comment (carmocca) [reply pytorch_lightning/callbacks/quantization.py:195]: I was referring to

`pl_module.dequant = torch.quantization.DeQuantStub()`

File: pytorch_lightning/callbacks/quantization.py (hunk 6 above)
Comment (carmocca) [reply pytorch_lightning/callbacks/quantization.py:193]: > Maybe a user warning would be good though?

Sure

File: pytorch_lightning/callbacks/quantization.py (hunk 5 above)
Comment (addisonklinke) [reply pytorch_lightning/callbacks/quantization.py:195]: Ah I think that's fine without a list. The forum [post](https://discuss.pytorch.org/t/is-it-possible-to-quantize-model-have-more-than-one-input/115018/2) I'm going off showed that you only need unique `QuantStub` elements, but the `DeQuantStub` can be shared globally by all inputs

File: pytorch_lightning/callbacks/quantization.py [hunk 7]
```diff
@@ -63,9 +66,9 @@ def wrap_quantize_forward_context(model: 'pl.LightningModule', func: Callable) -
     """
     # todo: consider using registering hook before/after forward
     @functools.wraps(func)
-    def wrapper(data) -> Any:
-        data = model.quant(data)
-        data = func(data)
+    def wrapper(*data, **kwargs) -> Any:
+        data = [q(d) for q, d in zip(model.quants, data)]
+        data = func(*data, **kwargs)
         data = model.dequant(data)
```
Comment (tchaton) [inline pytorch_lightning/callbacks/quantization.py:72]: Can we add multi-output support by the same occasion ?

**Review by tchaton — COMMENTED** [review]
Overall looks good.

File: pytorch_lightning/callbacks/quantization.py (hunk 7 above)
Comment (addisonklinke) [reply pytorch_lightning/callbacks/quantization.py:72]: We can re-use a single dequant stub for multiple outputs (since it's [stateless](https://discuss.pytorch.org/t/what-do-de-quantstub-actually-do/82402/2)), so yes it would just be a matter of iterating over the outputs

File: pytorch_lightning/callbacks/quantization.py [hunk 8]
```diff
@@ -114,6 +118,11 @@ def __init__(
             observer_type: allows switching between ``MovingAverageMinMaxObserver`` as "average" (default)
                 and ``HistogramObserver`` as "histogram" which is more computationally expensive.
 
+            max_num_inputs: maximum number of inputs nodes used by the model in training, evaluation, or
```
Comment (carmocca) [inline pytorch_lightning/callbacks/quantization.py:121]: Can you be more explicit about what are inputs exactly?

Is it only batches? I see you used `len(tensor)` to get the first dimension.

Same applies to other mentions of "inputs" in this PR

**Review by carmocca — COMMENTED** [review]
Tests are failing 🔍

File: pytorch_lightning/callbacks/quantization.py (hunk 7 above)
Comment (addisonklinke) [reply pytorch_lightning/callbacks/quantization.py:72]: I realized if we do `data = [model.dequant(d) for d in data]`, then we must make an opinionated choice about whether the final `data` is a list, tuple, or tensor. Since we don't know the expected return type of a users model, I don't see a good way to guarantee that the wrapper is consistent with their default forward behavior. Do you think we should just pick one of the iterable types and note it in the docs?

File: pytorch_lightning/callbacks/quantization.py (hunk 8 above)
Comment (addisonklinke) [reply pytorch_lightning/callbacks/quantization.py]: Do these latest two commits add enough clarity in this area?

File: pytorch_lightning/callbacks/quantization.py [hunk 9]
```diff
@@ -174,7 +208,21 @@ def _check_feasible_fuse(self, model):
...
+            if pl_module.example_input_array is None:
+                num_inps = 1
+                rank_zero_warn(
+                    'Defaulting to singular QuantStub. Use `max_num_inputs` or `model.example_input_array` '
+                    'if your model takes multiple inputs', UserWarning
+                )
+            else:
+                input_sample = pl_module.example_input_array
+                if not isinstance(input_sample, (tuple, list)):
+                    input_sample = (input_sample, )
```
Comment (carmocca) [inline pytorch_lightning/callbacks/quantization.py:221]: Why is this necessary?

File: pytorch_lightning/callbacks/quantization.py (hunk 7 above)
Comment (carmocca) [reply pytorch_lightning/callbacks/quantization.py:72]: What about using the type of the initial `data`?

Comment (addisonklinke) [reply pytorch_lightning/callbacks/quantization.py:72]: Yes @carmocca we could check the initial type of data returned from `func`, that's a good solution. I don't think this is necessary for the `QuantStub` application prior to `func` because the `*data` invocation is going to unpack the iterable regardless of what exact type it is

File: pytorch_lightning/callbacks/quantization.py (hunk 9 above)
Comment (addisonklinke) [reply pytorch_lightning/callbacks/quantization.py:221]: I suppose it's not strictly necessary since `len(torch.Tensor) == len(tuple(t))`. I thought it might make that behavior more readable in light of the docstring changes, but can remove if you'd like

**Comment by addisonklinke** [comment]
> Tests are failing

I checked the details of the first one "PyTorch & Conda / conda (3.7, 1.5) (pull_request)" and some quantization tests are failing despite passing locally in my dev environment. Is the test name saying that I need to run with Python 3.7 / PyTorch 1.5? Trying to figure out how I can reproduce the failure locally in order to fix it

**Comment by carmocca** [comment]
> Is the test name saying that I need to run with Python 3.7 / PyTorch 1.5? Trying to figure out how I can reproduce the failure locally in order to fix it

We run separate jobs with different version configurations. Seems like several pytorch versions are failing, not just 1.5. Also the GPU test job

**Comment by addisonklinke** [comment]
@carmocca I created a conda env with the requirements (following the contributing [guidelines](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md#testing)) + override for PyTorch 1.5

```
~/pytorch_lightning$ conda env create -n pytorch-1.5 python=3.7
~/pytorch_lightning$ conda activate pytorch-1.5
(pytorch-1.5) ~/pytorch_lightning$ python -m pip install ".[dev, examples]"
(pytorch-1.5) ~/pytorch_lightning$ python -m pip install pre-commit
(pytorch-1.5) ~/pytorch_lightning$ conda install -c pytorch pytorch==1.5.0 torchvision torchtext
```

The quantization tests also fail on the latest master (f1fa4c47), so I don't think the issues with 1.5 are caused by my PR

```
(pytorch-1.5) ~/pytorch_lightning$ python -m pytest --disable-warnings -v tests/callbacks/test_quantization.py
============================================================ short test summary info =============================================================
FAILED tests/callbacks/test_quantization.py::test_quantization[True-average] - RuntimeError: iter.device(arg).is_cuda() INTERNAL ASSERT FAILED ...
FAILED tests/callbacks/test_quantization.py::test_quantization[True-histogram] - pytorch_lightning.utilities.exceptions.MisconfigurationExcepti...
FAILED tests/callbacks/test_quantization.py::test_quantization[False-average] - RuntimeError: iter.device(arg).is_cuda() INTERNAL ASSERT FAILED...
FAILED tests/callbacks/test_quantization.py::test_quantization[False-histogram] - pytorch_lightning.utilities.exceptions.MisconfigurationExcept...
=================================================== 4 failed, 7 passed, 73 warnings in 11.13s ====================================================
```

The runtime error appears to be a known [bug](https://github.com/pytorch/pytorch/issues/32680) in PyTorch 1.5.0 - should this be marked as an expected failure then for both this PR and master? The other two failures are because `pytorch_lightning.utilities.imports._TORCH_LOWER_EQUAL_1_4` is returning true despite my environment having 1.5.0 installed. I checked the source code here and `_compare_version()` will always return true since it triggers a type error

```python
>>> import torch
>>> torch.__version__
'1.5.0'
>>> import importlib
>>> pkg = importlib.import_module('torch')
>>> from packaging.version import Version
>>> pkg_version = Version(pkg)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/addison/miniconda3/envs/pytorch-1.5/lib/python3.7/site-packages/packaging/version.py", line 296, in __init__
    match = self._regex.search(version)
TypeError: expected string or bytes-like object
```

With regard to the remaining PyTorch versions (1.6 and 1.7), I see those test failures are unique to the tests added by my PR, so I will work on investigating further. Is there a place I can view the full list of package versions used to construct these CI environments? For instance, "PyTorch & Conda / conda (3.7, 1.5)" is not completely descriptive

- Should I be using 1.5.0 or 1.5.1?
- What about the torchvision and torchtext versions? These have dependency on the PyTorch version, and if I don't install the proper ones tests can fail for reasons unrelated to Lightning
- Which particular python version? My conda defaults to 3.7.10, but I want to make sure this matches exactly what the CI server is using

Thanks for the information

**Comment by carmocca** [comment]
> should this be marked as an expected failure then for both this PR and master?

Yes

> _compare_version() will always return true since it triggers a type error

The snippet you shared is not correct. It should be `pkg_version = Version(pkg.__version__)` which is what we do here:

https://github.com/PyTorchLightning/pytorch-lightning/blob/9444a08d5607f3f243dae840e4016b353bfe92f6/pytorch_lightning/utilities/imports.py#L57

The issue is that

https://github.com/PyTorchLightning/pytorch-lightning/blob/9444a08d5607f3f243dae840e4016b353bfe92f6/pytorch_lightning/utilities/imports.py#L66

Should be

`_TORCH_LOWER_EQUAL_1_4 = _compare_version("torch", operator.lt, "1.5.0")`

> Is there a place I can view the full list of package versions used to construct these CI environments? For instance, "PyTorch & Conda / conda (3.7, 1.5)" is not completely descriptive

You can see all the installed dependencies inside each job logs, under the "Update dependencies" step:

Screenshot 2021-06-29 at 19 55 56

> Should I be using 1.5.0 or 1.5.1?

The latest patch release available, 1.5.1 in this case

**Comment by edenlightning** [comment]
@addisonklinke how is this going? would love to merge this for 1.4!

**Comment by addisonklinke** [comment]
@edenlightning Unfortunately I haven't had time to revisit recently with internal work taking priority. Last I recall, some of the failing tests seemed quite difficult to resolve because quantization has been in beta for all of the PyTorch versions in Lightning's CI list. I might have some time next Monday/Tuesday to look again, but I'm not too confident I'll be able to get the tests passing right away

**Comment by carmocca** [comment]
>Last I recall, some of the failing tests seemed quite difficult to resolve because quantization has been in beta for all of the PyTorch versions in Lightning's CI list.

Just as a heads up, we just dropped support for torch 1.4 and 1.5 for the upcoming Lightning 1.4 release. Not sure if this will help your failures 🤗

File: CHANGELOG.md [hunk 10]
```diff
@@ -338,6 +344,8 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Return the `default_root_dir` as the `log_dir` when the logger is a `LoggerCollection` ([#8187](https://github.com/PyTorchLightning/pytorch-lightning/pull/8187))
 
 
+
+
```
Comment (Borda) [inline CHANGELOG.md:348]:
```suggestion
```

File: pytorch_lightning/callbacks/quantization.py [hunk 11]
```diff
@@ -126,6 +137,41 @@ def custom_trigger_last(trainer):
         input_compatible: preserve quant/dequant layers. This allows to feat any input as to the original model,
             but break compatibility to torchscript and export with ``torch.save``.
 
+        max_num_inputs: maximum number of inputs nodes used by the model in training, evaluation, or
+            testing. This determines how many :class:`torch.quantization.QuantStub`s are needed. If
+            not specified, the model's :attr:~pytorch_lightning.core.lightning.LightningModule.example_input_array`
+            property must be defined
+
```
Comment (Borda) [inline pytorch_lightning/callbacks/quantization.py:144]:
```suggestion
```

**Comment by Borda** [comment]
@abefetterman mind check the failing tests? :rabbit:

**Comment by Borda** [comment]
@addisonklinke seems it is not compatible with PT < 1.7

**Comment by Borda** [comment]
@addisonklinke apology that it took so long, but still, could you please have look at the updated PR? :rabbit:
